library/core/src/num/nonzero.rs

```diff
@@ -34,6 +34,12 @@ macro_rules! nonzero_integers {
             /// use std::mem::size_of;
             #[doc = concat!("assert_eq!(size_of::<Option<core::num::", stringify!($Ty), ">>(), size_of::<", stringify!($Int), ">());")]
             /// ```
+            ///
+            /// # Layout
+            ///
+            #[doc = concat!("`", stringify!($Ty), "` has the same layout as `", stringify!($Int), "`, except that zero is not")]
+            #[doc = concat!("a valid instance of `", stringify!($ty), "`. `Option<", stringify!($Ty), ">` has the same layout")]
+            #[doc = concat!("and ABI as `", stringify!($Int), "`.")]
```

Comment:
Is this *ABI* a guarantee that's documented somewhere else right now?  Or is it a new promise?

Maybe this could start by just changing the "size" to "size and alignment" in the existing paragraph and code.  That much, at least, is very non-controversial promise AFAIK.

@rustbot author

Reply:
Given that it's [not yet well-defined](https://github.com/rust-lang/unsafe-code-guidelines/issues/304), I'm using "ABI" here to mean layout and bit validity. How would you feel if I just said "layout and bit validity" explicitly here? As you said, layout is a non-controversial promise. Bit validity is, in turn, a logical implication.

Specifically, an n-bit non-zero type can represent 2^n - 1 states, and so `Option<NonZeroXxx>` can represent 2^n states (the `None` variant adds one more possible state). This is the same number of states as exist in the corresponding n-bit primitive type. The layout (and thus the size) of `Option<NonZeroXxx>` is guaranteed to be the same as that of the corresponding primitive - namely n bits. Since there are 2^n possible states and n bits used to represent them, every one of the 2^n possible n-bit sequences must correspond to a valid state. The same is true by definition for the corresponding n-bit primitive. Thus, `Option<NonZeroXxx>` and the corresponding primitive have the same bit validity.

Reply:
@scottmcm Bump :) See also https://github.com/rust-lang/rust/pull/94786#issuecomment-1524074355.

Reply:
Oh, sorry for taking a million years here -- it was still waiting-on-author and thus not in my queue.

How about leaving the *header* as layout, but explicitly saying "size", "alignment", and "bit pattern" in the prose?  When the list isn't long, might as well just be extra-clear by saying all of them.

It looks like "size" is already used in <https://doc.rust-lang.org/std/option/index.html#representation>, for example.

Reply:
Actually, I just noticed that #94786 landed.  How do you want to proceed with this given that one happened?

Reply:
I've rebased on that. The wording change here is now much simpler - just:

> ...is guaranteed to be compatible with `Xxx`, including in FFI.

to:

> ...is guaranteed to have the same layout and bit validity as `Xxx`.

My intention is that the latter language implies the former, so it shouldn't be problematic to remove the bit about FFI. I also haven't seen the phrase "compatible with" used elsewhere in documentation relating to layout or bit validity, so my hope is that the latter language is more clear if you're already familiar with other such docs.

cc @joshtriplett since IIUC, the "compatible with" language was [your suggestion](https://github.com/rust-lang/rust/pull/94786#discussion_r882043136) in the original PR.